### PR TITLE
New Feature: fastq file reader/writer

### PR DESCRIPTION
**Pull request by mariehoffmann**
Add fastq file format reader and writer with test, test data, and Python bindings.

### CODE REVIEW

**Comment by cbielow** [comment]
nice addition and good code in general.
Some comments: the tabs will cause the checker to fail - please replace with spaces.

IMHO `struct FASTQEntry` should go into DATASTRUCTURES, not as a subclass in FORMAT.

Also I'd make the members independent of SeqAn structures, i.e. use OpenMS::String instead of seqan:...
But I'm happy to hear what others think.

What would be the first use-case for this? Maybe if all applications are seqan-heavy anyways, it makes sense to expose the seqan structures.

File: src/openms/source/FORMAT/FASTQFile.cpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+    std::fstream in(filename.c_str(), std::ios::binary | std::ios::in);
+    seqan::RecordReader<std::fstream, seqan::SinglePass<> > reader(in);
+    seqan::CharString id;
+    seqan::CharString seq;
+    seqan::CharString qual;
+    String::size_type position = String::npos;
+    Size size_read(0);
+	
+	while (!atEnd(reader))
```
Comment (hroest) [inline src/openms/source/FORMAT/FASTQFile.cpp]: whitespace issue

File: src/openms/source/FORMAT/FASTQFile.cpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+  FASTQFile::~FASTQFile()
+  {
+
+  }
+
+  
+  void FASTQFile::load(const String& filename, vector<FASTQEntry>& data)
+  {
+	  data.clear();
```
Comment (hroest) [inline src/openms/source/FORMAT/FASTQFile.cpp]: whitespace issue

File: src/pyOpenMS/pxds/FASTQFile.pxd [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+cdef extern from "<OpenMS/FORMAT/FASTQFile.h>" namespace "OpenMS::FASTQFile":
+
+    cdef cppclass FASTQEntry:
+        FASTQEntry() nogil except +
+        FASTQEntry(FASTQEntry) nogil except +
+
+        String identifier
+        String description
+        String sequence
+		String quality
```
Comment (hroest) [inline src/pyOpenMS/pxds/FASTQFile.pxd]: whitespace issue

File: src/openms/include/OpenMS/FORMAT/FASTQFile.h [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+	  in description and the text from the next line until the next linebreak is stored
+      in sequence. A newline starting with + (and optionally an identifier) is then 
+	  followed by a quality score of the same length as the sequence.
+    */
+    struct FASTQEntry
+    {
+      seqan::CharString identifier;
+      seqan::CharString description;
+      seqan::CharString sequence;
+	  seqan::CharString quality;
```
Comment (hroest) [inline src/openms/include/OpenMS/FORMAT/FASTQFile.h]: whitespace issue

**Comment by hroest** [comment]
Great, I like this addition a lot, it puts a nice and sane wrapper around seqan. 

However, I it might make sense to actually provide some additional functions to interpret the strings, especially the quality strings. Right now, its not that useful to get a quality string as a set of characters. There should at least be a helper function that converts these strings into actual Phred scores.

**Comment by hroest** [comment]
I also agree with Chris that it would be best to store the strings as something from either OpenMS or `std::` where I would actually prefer `std::string` to `OpenMS::String`. It seems that seqan charstring is a typedef http://docs.seqan.de/seqan/1.4.2-dox/global_typedef_CharString.html and is basically a `char *` underneath.

**Comment by mariehoffmann** [comment]
Ok, but the Google C++ guidelines you do not follow them completely, right? Like curly brackets are preferably placed in the next line and more than 80 characters are also OK. 
Another thing, wanted to use lambda expressions, but did not find a C++11 flag set. So, you do not require it yet?

**Comment by jpfeuffer** [comment]
Yes matching brackets on the same column should be preferred according to our conventions. I am not sure if we have a line limit but imho 80 chars is a good rule. You will probably find code here violating it but for new code it should be enforced.
And no we do not support C++11 yet. Discussion planned for the developer meeting. See #980

**Comment by mariehoffmann** [comment]
Most recent commit should fulfill requested changes. Please check again.

File: src/openms/include/OpenMS/DATASTRUCTURES/FASTQEntry.h [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  bool operator==(const FASTQEntry& rhs) const
+  {
+    return identifier == rhs.identifier
+           && description == rhs.description
+           && sequence == rhs.sequence
+           && quality == rhs.quality;
+  }
+  
+  // Illumina 1.8+ Phred+33 with score ranges ['!', 'J'] corresponding to [0, 41] 
+  std::vector<int> qual2phred()
```
Comment (cbielow) [inline src/openms/include/OpenMS/DATASTRUCTURES/FASTQEntry.h]: would `Byte` (Types.h) be enough here?

also `std::vector<int> qual2phred() const`!

File: src/openms/include/OpenMS/FORMAT/FASTQFile.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Timo Sachsenberg $
```
Comment (cbielow) [inline src/openms/include/OpenMS/FORMAT/FASTQFile.h]: @timosachsenberg maintainer? is this intended?

File: src/openms/source/FORMAT/FASTQFile.cpp [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    seqan::CharString qual;
+    String::size_type position = String::npos;
+    Size size_read(0);
+
+    while (!atEnd(reader))
+    {
+      if (readRecord(id, seq, qual, reader, seqan::Fastq()) != 0)
+      {
+        String msg;
+        if (data.empty())
```
Comment (cbielow) [inline src/openms/source/FORMAT/FASTQFile.cpp]: coding style. Please use:
```
if ... 
{
 ...
}
else
{
.
}
```

File: src/openms/include/OpenMS/FORMAT/FASTQFile.h (hunk 6 above)
Comment (timosachsenberg) [inline src/openms/include/OpenMS/FORMAT/FASTQFile.h]: Probably not. But you can add me as a placeholder if you have no intentions in maintaining the code.

File: src/openms/include/OpenMS/DATASTRUCTURES/FASTQEntry.h [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  {
+    std::vector<int> ps;
+    std::transform(begin(this->quality), end(this->quality), std::back_inserter(ps), std::bind2nd(std::minus<int>(), '!'));
+    return ps;
+  }
+  
+};
+
+#endif  // OPENMS_DATASTRUCTURES_FASTQENTRY_H
```
Comment (cbielow) [inline src/openms/include/OpenMS/DATASTRUCTURES/FASTQEntry.h]: missing newline (cppcheck will complain)

File: src/pyOpenMS/pxds/FASTQFile.pxd [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    cdef cppclass FASTQEntry:
+        FASTQEntry() nogil except +
+        FASTQEntry(FASTQEntry) nogil except +
+
+        String identifier
+        String description
+        String sequence
+        String quality
+
```
Comment (cbielow) [inline src/pyOpenMS/pxds/FASTQFile.pxd]: newline?

File: src/openms/source/FORMAT/FASTQFile.cpp [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+             filename + "'! " + msg +  " Please check the file!");
+      }
+
+      FASTQEntry newEntry;
+      newEntry.sequence = seq;
+
+      // handle id
+      String id_tmp = String(toCString(id)).trim();
+
+      position = id_tmp.find_first_of(" \v\t");
```
Comment (cbielow) [inline src/openms/source/FORMAT/FASTQFile.cpp]: could you add a comment, why " \v\t"?

**Comment by cbielow** [comment]
sorry for taking so long to review. I think we're about ready to merge. Nice PR

**Comment by timosachsenberg** [comment]
what is the state of this PR? it seemed pretty finished. Any comments?

**Comment by hroest** [comment]
@timosachsenberg @mariehoffmann @cbielow can we merge here? Any major roadblocks apart from style fixes?

**Comment by cbielow** [comment]
How to proceed here?
there are some codacy issues and some review comments.
BTW: c++11 is fully supported now, so you could use lambdas instead of bind2nd

**Comment by timosachsenberg** [comment]
if anyone wants to adopt this PR feel free to reopen

**Pull request closed without merging**